openra with extra mods

msteen opened the pull request:
###### Motivation for this change
This updates the existing OpenRA package with extra mods (thanks to @fusion809) and tries to share as much as possible between the code necessary to package an engine or an out-of-tree mod. This PR incorporates the feedback given on the previous PR (https://github.com/NixOS/nixpkgs/pull/53163) and replaces it.

cc: @veprbl @FRidh 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

msteen committed:
openra with extra mods

FRidh commented on pkgs/top-level/all-packages.nix:
```diff
@@ -20805,7 +20805,45 @@ in
   openmw-tes3mp = libsForQt5.callPackage ../games/openmw/tes3mp.nix { };
 
-  openra = callPackage ../games/openra { lua = lua5_1; };
+  inherit (
```
This is too much code for in `all-packages.nix`.

msteen replied:
I have moved that code into its own file, so now it is just a one liner in `all-packages.nix`.

FRidh reviewed (changes requested):
What was the motivation for opening a *second* PR?

msteen commented:
The reason for a second PR is that I am no longer just adding packages, but in this PR I am also replacing the existing `openra` package.

veprbl commented on pkgs/games/openra/builds.nix:
```diff
@@ -0,0 +1,34 @@
+let
+  f = import ./common.nix;
+  common = f (builtins.intersectAttrs (builtins.functionArgs f) pkgs // {
+    inherit (pkgs) stdenv;
```
This line can be dropped?

msteen replied:
Yes it can. It was necessary at some point, but due to refactoring it no longer is, I will remove it.

veprbl commented on pkgs/games/openra/builds.nix:
```diff
@@ -0,0 +1,34 @@
+in rec {
+  buildOpenRAEngine = { name ? null, version, description, homepage, mods, src, installExperimental ? "" }@engine:
+    let builder = name: pkgs.callPackage ./engine.nix (common // {
+      engine = engine // { inherit name installExperimental; };
```
Seems like here you override `engine.installExperimental` with `engine.installExperimental`

msteen replied:
Nope, that is not how destructering of attrs work in Nix:
```nix
(({ foo ? null }@attrs: attrs) {}).foo
```
`error: attribute 'foo' missing, at (string):1:1`
```nix
(({ foo ? null }@attrs: attrs // { inherit foo; }) {}).foo
```
`null`

veprbl commented on pkgs/games/openra/builds.nix:
```diff
@@ -0,0 +1,34 @@
+  buildOpenRAMod = { name ? null, version, title, description, homepage, src, engine }@mod: ({ version, mods ? [], src }@engine:
+    let builder = name: pkgs.callPackage ./mod.nix (common // {
+      mod = mod // { inherit name; };
+      engine = engine // { inherit mods; };
```
Seems like here you override `engine.mods` with `engine.mods`

veprbl commented on pkgs/games/openra/builds.nix:
```diff
@@ -0,0 +1,34 @@
+  common = f (builtins.intersectAttrs (builtins.functionArgs f) pkgs // {
+    inherit (pkgs) stdenv;
+    lua = pkgs.lua5_1;
+    inherit (pkgs.gnome3) zenity;
```
You don't need zenity to run the game?

msteen replied:
No, but it is nice to have a prompt show up when an error occurs rather than having to check the journal or running it from the command line to figure out why it is not running.

veprbl commented on pkgs/games/openra/mod.nix:
```diff
@@ -0,0 +1,86 @@
+
+  checkTarget = "test";
+
+  installPhase = ''
```
Add `runHook preInstall`

msteen replied:
Hmm, I have mixed feelings about this. It is not hard to work around not having these hooks present if you wanted to override something of the package and not all phases/hooks have to be available for a package, like when the `phases` attribute is used, but it is easy enough to add and I checked Nixpkgs and it is used 118 times, so I guess I will just add them.

msteen replied:
I am still not convinced, because you will have to check the package definition anyway to make sure you are not overwriting any existing `preInstall`. And since only few packages add those `runHook` calls, it is probably better to not assume adding such a phase will just work, instead it is best to always check the package definition. Is there some guideline stating they should be added? Any other arguments to change my mind? I would love to know, so far I have asked twice on the IRC, but gotten no response. However since you seem to feel strongly about it and it does not have any other effect on the build, so I have added them where possible.

veprbl replied:
A naive override of a phase without using `runHook` calls breaks pre- and post- hooks. This is not a consistent interface and can lead to unpleasant user experience. We could say that this is a bug in stdenv and we have to work around it. There is some work on improving this treewide: https://github.com/dezgeg/rfcs/blob/phase-changes/rfcs/0032-run-phase-changes-for-better-nix-shell-use.md

AFAIK there is no guideline for applying this workaround and you are right to point out the lack of consistency in its use across nixpkgs. And I do appreciate that it might make no sense for you to add it to your expression. I think, the only way to really resolve this would be to fix stdenv (e.g by means of RFC above).

FRidh replied:
We expect functions like `buildPythonPackage`, which may override phases, to ensure hooks are still available, but not every individual derivation. It's also difficult to do; what if the derivation implements a `postInstall`, should it then, when explicitly passed in a `postInstall`, override it, or extend it?

veprbl commented on pkgs/games/openra/mod.nix:
```diff
@@ -0,0 +1,86 @@
+      size=''${size}x''${size}
+      cp packaging/linux/mod_''${size}.png $(mkdirp $out/share/icons/hicolor/''${size}/apps)/openra-${mod.name}.png
+    done
+  '';
```
Add `runHook postInstall`

veprbl commented on pkgs/games/openra/mods.nix:
```diff
@@ -0,0 +1,246 @@
+{ buildOpenRAMod, fetchFromGitHub, extraPostFetch }:
+
+{
```
Probably `recurseIntoAttrs` is needed here

msteen replied:
This is not telling me much, could you elaborate?

FRidh replied:
`recurseIntoAttrs` is needed for `nix-env` and Hydra to find the derivations

veprbl commented on pkgs/games/openra/builds.nix:
```diff
@@ -0,0 +1,34 @@
+
+let
+  f = import ./common.nix;
+  common = f (builtins.intersectAttrs (builtins.functionArgs f) pkgs // {
```
Can't you just use `pkgs.callPackage`?

msteen replied:
Nope, I tried, but `callPackage` [makes it overridable](https://github.com/NixOS/nixpkgs/blob/c6e043d57c2cc69a1068a5e6e34b385942876d0e/lib/customisation.nix#L112), adding additional attributes to the returned attribute set, causing the argument destructering of `mod.nix` and `engine.nix` to fail due to the additional attribute names.

veprbl commented on pkgs/games/openra/mod.nix:
```diff
@@ -0,0 +1,86 @@
+    ${patchEngine engine.src.name engine.version}
+  '';
+
+  configurePhase = ''
```
Add `runHook preConfigure`

veprbl commented on pkgs/games/openra/mod.nix:
```diff
@@ -0,0 +1,86 @@
+  configurePhase = ''
+    make version VERSION=${escapeShellArg mod.version}
+    make -C ${engine.src.name} version VERSION=${escapeShellArg engine.version}
+  '';
```
Add `runHook postConfigure`

veprbl commented on pkgs/games/openra/mod.nix:
```diff
@@ -0,0 +1,86 @@
+  sourceRoot = ".";
+
+  postUnpack = ''
+    mv ${engine.src.name} ${mod.src.name}
```
So you can't build engine separately?

msteen replied:
I asked this too, but my own attempts failed. From what I understand the mods need to have the source of the engine available, and since every mod uses a specific engine so far, there is also not much gained with separately building the engine anyway.

fusion809 commented:
One worthy addition to this PR is the [Sole Survivor](https://github.com/MustaphaTR/sole-survivor) mod:

```bash
  ss = buildOpenRAMod {
    version = "72";
    title = "Sole Survivor";
    description = "A re-imagination of the original Command & Conquer: Sole Survivor game";
    homepage = https://github.com/MustaphaTR/sole-survivor;
    src = fetchFromGitHub {
      owner = "MustaphaTR";
      repo = "sole-survivor";
      rev = "fad65579c8b487cef9a8145e872390ed77c16c69";
      sha256 = "0h7is7x2qyvq7vqp0jgw5zrdkw8g7ndd82d843ldhnb0a3vyrk34";
    };
    engine = rec {
      version = "becfc15";
      src = fetchFromGitHub {
        owner = "OpenRA";
        repo = "OpenRA" ;
        rev = version;
        sha256 = "0id8vf3cjr7h5pz4sw8pdaz3sc45lxr21k1fk4309kixsrpa7i0y";
        name = "engine";
        inherit extraPostFetch;
      };
    };
  };
```

msteen commented:
I have added it to the mod list, thank you!

FRidh commented on pkgs/games/openra/builds.nix:
```diff
@@ -0,0 +1,33 @@
+    inherit (pkgs.gnome3) zenity;
+  });
+
+  buildOpenRASet = f: args: builtins.mapAttrs (name: value: if builtins.isFunction value then value name else value) (f ({
```
Document what `f` and `args` are.

msteen replied:
I have documented `common.nix` to explain that and more.

FRidh commented on pkgs/games/openra/engine.nix:
```diff
@@ -0,0 +1,39 @@
+with stdenv.lib;
+
+stdenv.mkDerivation (recursiveUpdate packageAttrs {
+  name = "${engine.name}-${engine.version}";
```
```
pname = engine.name;
version = engine.version;
```

msteen replied:
I have gotten different feedback in the past, that it is better to keep those in their own let, instead of making them part of the derivation attributes, and if so then you do not gain anything (putting `version = engine.version` in a let, that is). So this is a bit of contradictory feedback to me, could you give any good arguments as to why I should ignore my previous feedback over yours?

FRidh replied:
Correct, in the past it was undesirable because these weren't parameters to `stdenv.mkDerivation`, however, `pname` and `version` are nowadays parameters of `stdenv.mkDerivation`.

msteen replied:
Adding `name` should not hurt, so I have just added it explicitly as well.

FRidh commented on pkgs/top-level/all-packages.nix:
```diff
@@ -20805,7 +20805,9 @@ in
   openmw-tes3mp = libsForQt5.callPackage ../games/openmw/tes3mp.nix { };
 
-  openra = callPackage ../games/openra { lua = lua5_1; };
+  inherit (import ../games/openra/builds.nix pkgs) buildOpenRAEngine buildOpenRAMod openra-engines openra-mods;
```
please keep this in its own namespace

msteen replied:
Done.

FRidh reviewed (changes requested):
How does one use the engines and mods?

msteen commented:
> How does one use the engines and mods?

Both engines and mods provide you with OpenRA mods to play. The difference is that an engine provides several mods as part of the engine, and out-of-tree (or out-of-engine if you will) mods provide only that particular mod.

For example `nix-build '<nixpkgs>' --attr openraPackages.engines.stable` provides `./result/bin/openra-{cnc,d2k,ra,ts}`, and `nix-build '<nixpkgs>' --attr openraPackages.mods.ss` provides just `./result/bin/openra-ss`

Even though I do not know of any engine forks that provide other mods as part of the engine, it was easy to define engines like mods in a generic way and it would for example now also allow for a bleeding edge version of the OpenRA engine to exist.

The out-of-tree mods require the source code of the engine as part of their building process. And the engine needed is always some specific version of OpenRA or some fork of it.

FRidh commented:
Could you describe this in the main `default.nix`. Basically, give a brief overview of the various files, what they're for, and how one actually uses it.

msteen commented:
There is no `default.nix`, but I guess I can place a comment, describing what they do, above `common.nix`, 'engine.nix', and `mod.nix`.

FRidh reviewed (commented):
This looks good to me.

One more small suggestion. In the top of `packages.nix`, state that this file contains all `openra` packages, and that e.g. Sole Survivor is the attribute `openraPackages.mods.ss`.

FRidh commented:
> There is no default.nix

Maybe move `packages.nix` to `default.nix` then? It's typically the name of the first file one goes into.

msteen commented:
I have implemented all of your feedback.

The `packages.nix` file has been renamed to `default.nix`, explaining that it defines all OpenRA packages, and I have added a bit of an overview to it.

I have added `recurseIntoAttrs` to the `buildOpenRASet` function, so both the engine and mod sets will have them set.

FRidh reviewed (approved):
Thanks!

fusion809 commented on pkgs/games/openra/mods.nix:
```diff
@@ -0,0 +1,270 @@
+
+  kknd = buildOpenRAMod {
+    version = "136";
+    title = "Krush, Kill n' Destroy";
```
I've been looking through the web and it seems like the comma (,) in this title (and hence also in the line below) is not common place, so maybe it should be removed from here.

veprbl commented on pkgs/games/openra/default.nix:
```diff
@@ -1,79 +1,70 @@
+  } // args)));
 
-  installTargets = [ "install" "install-linux-icons" "install-linux-desktop" "install-linux-appdata" "install-linux-mime" "install-man-page" ];
+in rec {
```
This needs `recurseIntoAttrs` too

veprbl replied:
You can test if it works locally: `nix-env -qa ".*openra.*" -f .` should show all your mods and engines.

msteen replied:
Right, since they are now in their own namespace `openraPackages`, that attrset should also be marked with `recurseIntoAttrs` in order for `nix-env` to see them. I tested it and they can now be found, thank you!

veprbl commented:
This one fails to build for me:
@GrahamcOfBorg build openraPackages.mods.yr

veprbl commented:
This one too:
@GrahamcOfBorg build openraPackages.mods.rv

veprbl commented:
This too?
@GrahamcOfBorg build openraPackages.mods.d2

veprbl commented:
@GrahamcOfBorg openraPackages.mods.sp

fusion809 commented:
I think we can ignore them, because all four of the mods with test failures, you flagged, run to fine for me, after being built without tests. I've started single player campaigns, created a few units, infiltrated some buildings with my engineers, _etc._ 

Beware, if you try testing them yourself, D2 crashes on start up, with an error message complaining of a missing file called `IBM.PAL`, if you do not have the required game assets. If you do have the appropriate files in ~/.openra/Content/d2, on the other hand, it launches and runs fine.

veprbl commented:
@GrahamcOfBorg build openra
@GrahamcOfBorg build openraPackages.engines.bleed
@GrahamcOfBorg build openraPackages.engines.playtest
@GrahamcOfBorg build openraPackages.mods.ca
@GrahamcOfBorg build openraPackages.mods.d2
@GrahamcOfBorg build openraPackages.mods.dr
@GrahamcOfBorg build openraPackages.mods.gen
@GrahamcOfBorg build openraPackages.mods.kknd
@GrahamcOfBorg build openraPackages.mods.mw
@GrahamcOfBorg build openraPackages.mods.ra2
@GrahamcOfBorg build openraPackages.mods.raclassic
@GrahamcOfBorg build openraPackages.mods.rv
@GrahamcOfBorg build openraPackages.mods.sp
@GrahamcOfBorg build openraPackages.mods.ss
@GrahamcOfBorg build openraPackages.mods.ura
@GrahamcOfBorg build openraPackages.mods.yr

msteen commented on pkgs/games/openra/default.nix:
```diff
@@ -1,79 +1,70 @@
+      if the attribute name and engine/mod name are equal.
+  */
+  callWithName = name: value: if builtins.isFunction value then value name else value;
+  buildOpenRASet = f: args: pkgs.recurseIntoAttrs (builtins.mapAttrs callWithName (f ({
```
Though it should be safe for usage in 18.09 and up, right? Anyway, I have changed it to use `lib` instead as you suggested.

veprbl replied:
Yes. Judging by release notes you are right about NixOS version. However, nixpkgs must evaluate on Nix as old as 2.0 as stated by current https://github.com/NixOS/nixpkgs/blob/master/lib/minver.nix

msteen commented:
I have build them all locally, I had to create a `unsafeBuildOpenRAMod` that disables checking for 4 mods that fail with errors when testing the maps, but seem to work without error when just playing the game, as as tested by @fusion809.

Some mods need unfree assets to be installed, so instead of letting them crash with an error message or letting them ask to look for the installation media of the original games, I have added an error message stating that they need those assets with a link for more information.

msteen commented:
I forgot the push changing the name/version of the engines. I thought it better to move the engine name part in the version, since whether it is release, playtest or bleed is saying something about the OpenRA version, not the name.

@veprbl I am sorry this push undid the results of the requested builds, but I checked before pushing and they all passed except for the Darwin builds, which where neutral since they did not apply.
